pkg/controller/kms/cryptokey.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-gcp/apis/kms/v1alpha1"
+	gcp "github.com/crossplane/provider-gcp/pkg/clients"
+	"github.com/crossplane/provider-gcp/pkg/clients/cryptokey"
+)
+
+const (
+	errNotCryptoKey        = "managed resource is not a GCP CryptoKey"
+	errManagedUpdateFailed = "cannot update CryptoKey custom resource"
+	errCheckUpToDate       = "cannot determine if CryptoKey instance is up to date"
+)
+
+// SetupCryptoKey adds a controller that reconciles CryptoKeys.
+func SetupCryptoKey(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.CryptoKeyGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.CryptoKey{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.CryptoKeyGroupVersionKind),
+			managed.WithExternalConnecter(&cryptoKeyConnecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type cryptoKeyConnecter struct {
+	client client.Client
+}
+
+// Connect sets up kms client using credentials from the provider
+func (c *cryptoKeyConnecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	_, opts, err := gcp.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+	s, err := kmsv1.NewService(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &cryptoKeyExternal{kube: c.client, cryptokeys: kmsv1.NewProjectsLocationsKeyRingsCryptoKeysService(s)}, nil
+}
+
+type cryptoKeyExternal struct {
+	kube       client.Client
+	cryptokeys cryptokey.Client
+}
+
+func (e *cryptoKeyExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.CryptoKey)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotCryptoKey)
+	}
+
+	// Hack to cleanup CR without deleting actual resource.
+	// It is not possible to delete KMS CryptoKeys, there is no "delete" method defined:
+	// https://cloud.google.com/kms/docs/reference/rest#rest-resource:-v1.projects.locations.keyrings.cryptokeys
+	// Also see related faq: https://cloud.google.com/kms/docs/faq#cannot_delete
+	if meta.WasDeleted(cr) {
+		return managed.ExternalObservation{
+			ResourceExists:    false,
+			ResourceUpToDate:  false,
+			ConnectionDetails: managed.ConnectionDetails{},
+		}, nil
+	}
+
+	instance, err := e.cryptokeys.Get(cryptoKeyRRN(cr)).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGet)
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	cryptokey.LateInitializeSpec(&cr.Spec.ForProvider, *instance)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
```

Comment:
In new version of crossplane-runtime, we have `ResourceLateInitialized` field that you can use. If it's `true`, we'll issue a `kube.Update` for you.

Reply:
Ah, good to know. Done!